pallets/pips/src/lib.rs

```diff
@@ -809,194 +864,264 @@ decl_module! {
-            Self::create_referendum(
-                id,
-                ReferendumState::Pending,
-                ReferendumType::Emergency,
-            );
+        pub fn clear_snapshot(origin) -> DispatchResult {
+            // 1. Check that a GC member is executing this.
+            let actor = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&actor)?;
+            ensure!(T::GovernanceCommittee::is_member(&did), Error::<T>::NotACommitteeMember);
+
+            // 2. Clear the snapshot.
+            <SnapshotMeta<T>>::kill();
+            <SnapshotQueue<T>>::kill();
+
+            // 3. Emit event.
+            Self::deposit_event(RawEvent::SnapshotCleared(did));
             Ok(())
         }
 
-        /// Moves a referendum instance into dispatch queue.
-        #[weight = (800_000_000, DispatchClass::Operational, Pays::Yes)]
-        pub fn enact_referendum(origin, id: PipId) -> DispatchResult {
-            T::VotingMajorityOrigin::ensure_origin(origin)?;
-            // Check that referendum is Pending
-            Self::is_referendum_state(id, ReferendumState::Pending)?;
-            Self::prepare_to_dispatch(id)
-        }
+        /// Takes a new snapshot of the current list of active && pending PIPs.
+        /// The PIPs are then sorted into a priority queue based on each PIP's weight.
+        ///
+        /// # Errors
+        /// * `NotACommitteeMember` - triggered when a non-GC-member executes the function.
+        #[weight = (1_000_000_000, DispatchClass::Operational, Pays::Yes)]
+        pub fn snapshot(origin) -> DispatchResult {
+            // 1. Check that a GC member is executing this.
+            let made_by = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&made_by)?;
+            ensure!(T::GovernanceCommittee::is_member(&did), Error::<T>::NotACommitteeMember);
+
+            // 2. Fetch intersection of pending && non-cooling PIPs and aggregate their votes.
+            let created_at = <system::Module<T>>::block_number();
+            let mut queue = <Proposals<T>>::iter_values()
+                // Only keep pending PIPs.
+                .filter(|pip| matches!(pip.state, ProposalState::Pending))
+                .map(|pip| pip.id)
+                // Only keep community PIPs not cooling-off.
+                .filter(|id| {
+                    <ProposalMetadata<T>>::get(id)
+                        .filter(|meta| meta.cool_off_until <= created_at)
+                        .filter(|meta| matches!(meta.proposer, Proposer::Community(_)))
+                        .is_some()
+                })
+                // Aggregate the votes; `true` denotes a positive sign.
+                .map(|id| {
+                    let VotingResult { ayes_stake, nays_stake, .. } = <ProposalResult<T>>::get(id);
+                    let weight = if ayes_stake >= nays_stake {
+                        (true, ayes_stake - nays_stake)
```

Comment:
Ideally we would have a single weight (rather than a (bool, weight) tuple) for each PIP - this seems more general (i.e. the bool bit may not make sense with a different weighting algorithm).

We could scale linearly using something like:
weight = x - min(X) / max(X) - min(X)